cartographer/mapping/internal/3d/pose_graph_3d.cc

```diff
@@ -256,111 +254,132 @@ void PoseGraph3D::ComputeConstraint(const NodeId& node_id,
   const transform::Rigid3d global_submap_pose_inverse =
       global_submap_pose.inverse();
 
+  bool maybe_add_local_constraint = false;
+  bool maybe_add_global_constraint = false;
+  const TrajectoryNode::Data* constant_data;
+  const Submap3D* submap;
   std::vector<TrajectoryNode> submap_nodes;
-  for (const NodeId& submap_node_id :
+  {
+    common::MutexLocker locker(&mutex_);
+    CHECK(data_.submap_data.at(submap_id).state == SubmapState::kFinished);
+
+    for (const NodeId& submap_node_id :
        data_.submap_data.at(submap_id).node_ids) {
-    submap_nodes.push_back(TrajectoryNode{
-        data_.trajectory_nodes.at(submap_node_id).constant_data,
-        global_submap_pose_inverse *
-            data_.trajectory_nodes.at(submap_node_id).global_pose});
+      submap_nodes.push_back(TrajectoryNode{
+          data_.trajectory_nodes.at(submap_node_id).constant_data,
+          global_submap_pose_inverse *
+              data_.trajectory_nodes.at(submap_node_id).global_pose});
+    }
+
+    const common::Time node_time = GetLatestNodeTime(node_id, submap_id);
+    const common::Time last_connection_time =
+        data_.trajectory_connectivity_state.LastConnectionTime(
+            node_id.trajectory_id, submap_id.trajectory_id);
+    if (node_id.trajectory_id == submap_id.trajectory_id ||
+        node_time <
+            last_connection_time +
+                common::FromSeconds(
+                    options_.global_constraint_search_after_n_seconds())) {
+      maybe_add_local_constraint = true;
+    } else if (
+        global_localization_samplers_[node_id.trajectory_id]->Pulse()) {
+      maybe_add_global_constraint = true;
+    }
+    constant_data = data_.trajectory_nodes.at(node_id).constant_data.get();
+    submap = static_cast<const Submap3D*>(
+        data_.submap_data.at(submap_id).submap.get());
   }
 
-  const common::Time node_time = GetLatestNodeTime(node_id, submap_id);
-  const common::Time last_connection_time =
-      data_.trajectory_connectivity_state.LastConnectionTime(
-          node_id.trajectory_id, submap_id.trajectory_id);
-  if (node_id.trajectory_id == submap_id.trajectory_id ||
-      node_time <
-          last_connection_time +
-              common::FromSeconds(
-                  options_.global_constraint_search_after_n_seconds())) {
+  if (maybe_add_local_constraint) {
     // If the node and the submap belong to the same trajectory or if there has
```

Comment:
Can we move this up where the bool gets set? The comment is much easier to understand next to the if statement. Similar for the comment below.